### PR TITLE
Refactor most of the API endpoints

### PR DESCRIPTION
**Pull request by Weredime**

### CODE REVIEW

**Comment by Looky1173** [comment]
@Weredime Thank you for the PR! Everything looks fine and works great in the preview, but I'm worried about the code duplication that splitting the jam API introduced.

**Comment by Weredime** [comment]
@Looky1173 As the server is hosted on vercel, it wouldn't make a difference as vercel is serverless - It has to create a new db connection for every request, so there won't be much of a difference.
On the development side though, there is one database connection so it's ok

I'll try to refactor it a bit more to add simple functions like isAdmin.

**Comment by Looky1173** [comment]
@Weredime Got it; thank you again for your contribution!

File: src/pages/api/scratch-jams/[jam]/hosts/index.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    switch (req.method) {
+        case 'GET': {
+            const hosts = (await getScratchGameJamHosts(Database, jam)) || [];
+            return res.status(200).json({ hosts: hosts, isOrganizer: isOrganizer });
+        }
+        case 'POST': {
+            let body = req.body;
+            if (isObjectEmpty(body)) return res.status(400).json({ error: { identifier: 'missingParameters', message: 'Missing body with host username' } });
+            if (!isObject(body)) body = JSON.parse(body);
```
Comment (Looky1173) [inline src/pages/api/scratch-jams/[jam]/hosts/index.js:24]: `isObjectEmpty` and `isObject` are not defined in the current scope.

**Review by Looky1173 — DISMISSED** [review]
The hosts API is broken:
image

File: src/pages/api/scratch-jams/[jam]/hosts/index.js (hunk 1 above)
Comment (Looky1173) [reply src/pages/api/scratch-jams/[jam]/hosts/index.js:24]: I edited your PR so this comment is no longer relevant.

**Comment by Looky1173** [comment]
@Weredime Sorry, this isn't going to work -- there are just too many errors and regressions. I will close this pull request and gradually cherry-pick and integrate your changes that do function.

**Comment by Weredime** [comment]
I'll try to fix this, and if its alright we can reopen?

**Comment by Looky1173** [comment]
> I'll try to fix this, and if its alright we can reopen?

@Weredime That wold be great! Also, could you please split your PR into two, with one containing the minor code formatting edits made with Prettier, and the other being the refactor of the API.

Finally, don't forget about the following that must be included for all routes:
```js
const jam = await getScratchGameJam(Database, slug[0], req.session?.user?.name);
    if (jam.meta?.archived === true && !isAdmin)
        return res.status(403).json({ error: { identifier: 'insufficientPermissions', message: 'You do not have permission to access archived game jams. Only administrators can do that.' } });
```

```js
if ((req.method === 'POST' || req.method === 'PUT' || req.method === 'DELETE') && !req.session.user)
        return res.status(401).json({ error: { identifier: 'authorizationRequired', message: 'Authorization is required for this endpoint' } });
```

**Comment by Looky1173** [comment]
Maybe you could isolate the above code into middlewares like how `iron-session` does it?

**Comment by Weredime** [comment]
Might as well

**Comment by Weredime** [comment]
I'll refactor how the database functions work so it'd be much easier, and maybe add a middleware for db?

**Comment by Looky1173** [comment]
That sounds good!

**Comment by Looky1173** [comment]
@Weredime I think the middleware could be done like so:
```js
import { withIronSessionApiRoute } from 'iron-session/next';
import { sessionOptions } from '@constants';

const ensureArchivedJamIntegrity = (req, res, params) => {
  const jam = await getScratchGameJam(Database, params.jam, req.session?.user?.name);
  if (jam.meta?.archived === true && !isAdmin)
    return res.status(403).json({ error: { identifier: 'insufficientPermissions', message: 'You do not have permission to access archived game jams. Only administrators can do that.' } });
}

// Helper method to wait for a middleware to execute before continuing
// And to throw an error when an error happens in a middleware
function runMiddleware(req, res, params, fn) {
  return new Promise((resolve, reject) => {
    fn(req, res, params, (result) => {
      if (result instanceof Error) {
        return reject(result);
      }

      return resolve(result);
    })
  })
}

async function route(req, res) {
  const { jam } = req.query;

  // Run the middleware
  await runMiddleware(req, res, { jam: jam }, ensureArchivedJamIntegrity );

  // Rest of the API logic
  res.status(200)json({ success: true });
}

export default route;
```

**Pull request closed without merging**